lib/gc/Transforms/GPU/GpuToGpuOcl.cpp

```diff
@@ -0,0 +1,532 @@
+//===-- GpuToGpuOcl.cpp - GpuToGpuOcl path ----------------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#include <unordered_set>
+
+#define GC_GPU_OCL_CONST_ONLY
+#include "gc/ExecutionEngine/GPURuntime/GpuOclRuntime.h"
+
+#include "mlir/Conversion/LLVMCommon/ConversionTarget.h"
+#include "mlir/Conversion/LLVMCommon/Pattern.h"
+#include "mlir/Conversion/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/Transforms/Passes.h"
+
+using namespace mlir;
+using namespace mlir::gc::gpu;
+
+namespace mlir::gc {
+#define GEN_PASS_DECL_GPUTOGPUOCL
+#define GEN_PASS_DEF_GPUTOGPUOCL
+#include "gc/Transforms/Passes.h.inc"
+} // namespace mlir::gc
+
+namespace {
+LLVM::CallOp funcCall(OpBuilder &builder, const StringRef name,
+                      const Type returnType, const ArrayRef<Type> argTypes,
+                      const Location loc, const ArrayRef<Value> arguments,
+                      bool isVarArg = false) {
+  auto module = builder.getBlock()->getParent()->getParentOfType<ModuleOp>();
+  auto function = module.lookupSymbol<LLVM::LLVMFuncOp>(name);
+  if (!function) {
+    auto type = LLVM::LLVMFunctionType::get(returnType, argTypes, isVarArg);
+    function = OpBuilder::atBlockEnd(module.getBody())
+                   .create<LLVM::LLVMFuncOp>(loc, name, type);
+  }
+  return builder.create<LLVM::CallOp>(loc, function, arguments);
+}
+
+// Assuming that the pointer to the context is passed as the last argument
+// of the current function of type memref<anyType> with zero dims. When lowering
```

Comment:
A new argument is added to the initial function, before running the passes. We can add a single pointer, but:
1. in this case we need to also change the function call in _mlir_ciface_entry(), that is created by the pipeline.
2. it does not make sense, because all args to the function are passed through the function wrapper and are wrapped into the https://mlir.llvm.org/docs/TargetLLVMIR/#c-compatible-wrapper-emission structure, that has minimum 3 fields.